ruler/ruler.go

```diff
@@ -17,218 +15,172 @@ import (
+	for i := 0; i < cfg.NumWorkers; i++ {
+		workers[i] = newWorker(&s, ruler)
 	}
-	return &Ruler{
-		cfg:           cfg,
-		chunkStore:    chunkStore,
-		distributor:   d,
-		configsAPIURL: configsAPIURL,
-		externalURL:   externalURL,
+	return &Server{
+		scheduler: &s,
+		workers:   workers,
 	}, nil
 }
 
-// GetWorkerFor gets a rules recording worker for the given user.
-// It will keep polling until it can construct one.
-func (r *Ruler) GetWorkerFor(userID string) Worker {
-	delay := time.Duration(r.cfg.EvaluationInterval)
-	return &worker{
-		delay:         delay,
-		userID:        userID,
-		configsAPIURL: r.configsAPIURL,
-		opts:          r.getManagerOptions(userID),
+// Run the server.
+func (s *Server) Run() {
+	go s.scheduler.Run()
+	for _, w := range s.workers {
+		go w.Run()
 	}
+	log.Infof("Ruler up and running")
 }
 
-func (r *Ruler) getManagerOptions(userID string) *rules.ManagerOptions {
-	ctx := user.WithID(context.Background(), userID)
-	appender := appenderAdapter{distributor: r.distributor, ctx: ctx}
-	queryable := querier.NewQueryable(r.distributor, r.chunkStore)
-	engine := promql.NewEngine(queryable, nil)
-	return &rules.ManagerOptions{
-		SampleAppender: appender,
-		Notifier:       nil,
-		QueryEngine:    engine,
-		Context:        ctx,
-		ExternalURL:    r.externalURL,
+// Stop the server.
+func (s *Server) Stop() {
+	for _, w := range s.workers {
+		w.Stop()
 	}
+	s.scheduler.Stop()
 }
 
-// loadRules loads rules.
-//
-// Strongly inspired by `loadGroups` in Prometheus.
-func loadRules(files map[string]string) ([]rules.Rule, error) {
-	result := []rules.Rule{}
-	for fn, content := range files {
-		stmts, err := promql.ParseStmts(string(content))
-		if err != nil {
-			return nil, fmt.Errorf("error parsing %s: %s", fn, err)
-		}
-
-		for _, stmt := range stmts {
-			var rule rules.Rule
-
-			switch r := stmt.(type) {
-			case *promql.AlertStmt:
-				rule = rules.NewAlertingRule(r.Name, r.Expr, r.Duration, r.Labels, r.Annotations)
-
-			case *promql.RecordStmt:
-				rule = rules.NewRecordingRule(r.Name, r.Expr, r.Labels)
-
-			default:
-				panic("ruler.loadRules: unknown statement type")
-			}
-			result = append(result, rule)
-		}
-	}
-	return result, nil
+// Worker does a thing until it's told to stop.
+type Worker interface {
+	Run()
+	Stop()
 }
 
-// appenderAdapter adapts a distributor.Distributor to prometheus.SampleAppender
-type appenderAdapter struct {
-	distributor *distributor.Distributor
-	ctx         context.Context
-}
+type worker struct {
+	scheduler *scheduler
+	ruler     Ruler
 
-func (a appenderAdapter) Append(sample *model.Sample) error {
-	req := util.ToWriteRequest([]*model.Sample{sample})
-	_, err := a.distributor.Push(a.ctx, req)
-	return err
+	done       chan struct{}
+	terminated chan struct{}
 }
 
-func (a appenderAdapter) NeedsThrottling() bool {
-	// XXX: Just a guess. Who knows?
-	return false
+func newWorker(scheduler *scheduler, ruler Ruler) worker {
```

Comment:
Actually, looking at the type, I think the constructor *should* do more. Shouldn't it `make` the `done` and `terminated` channels? Otherwise they'll be `nil` and will cause crashes or block.